File: backend-laravel/app/Models/PermisoDiario.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

/**
 * Permiso diario para tomar asistencia (tabla `permisos_diarios`). La
 * apertura es manual —el jefe de preceptores abre el día explícitamente,
 * `usuario_apertura_id` es obligatorio tal como pide la narrativa— pero
 * el cierre NO se marca a mano ni depende de ningún proceso programado:
 * se calcula comparando la hora actual contra `hora_limite` (23:59:59
 * por defecto). `cerrado_manual` solo existe para un cierre anticipado
 * explícito si hiciera falta.
 *
 * La automatización completa de la apertura (a una hora fija) se evaluó,
 * se implementó y se probó, y finalmente se revirtió
 *  — ver `Documentacion_Base_de_Datos.pdf`, sección 6.5.
 * No es un olvido, es la decisión final.
 *
 * El propio trigger `trg_planillas_before_insert` en MySQL ya impide
 * crear una planilla de HOY sin este permiso abierto y vigente —
 * `estaVigenteHoy()` acá abajo es la misma regla del lado de PHP, para
 * poder validar (y dar un mensaje claro) antes de llegar a la base.
 */
class PermisoDiario extends Model
{
    protected $table = 'permisos_diarios';

    protected $primaryKey = 'id_permiso_diario';

    protected $fillable = [
        'fecha',
        'usuario_apertura_id',
        'hora_apertura',
        'hora_limite',
        'cerrado_manual',
    ];

    protected function casts(): array
    {
        return [
            'fecha' => 'date',
            'hora_apertura' => 'datetime',
            'cerrado_manual' => 'boolean',
        ];
    }

    public function usuarioApertura(): BelongsTo
    {
        return $this->belongsTo(Usuario::class, 'usuario_apertura_id', 'id_usuario');
    }

    /**
     * Espejo en PHP de `vista_permisos_diarios_vigentes`: ¿hay un
     * permiso de HOY, no cerrado a mano, y todavía no pasó la hora
     * límite?
     */
    public static function estaVigenteHoy(): bool
    {
        return static::query()
            ->whereDate('fecha', now()->toDateString())
            ->where('cerrado_manual', false)
            ->whereTime('hora_limite', '>=', now()->format('H:i:s'))
            ->exists();
    }
}

File: app_flutter/lib/models/permiso_diario.dart
/// Espejo de lo que devuelve `PermisosDiariosController::formatear()` en
/// el backend (`GET/POST/PATCH /permisos-diarios/...`) — el permiso para
/// tomar asistencia de un día puntual. Ver `App\Models\PermisoDiario`
/// para las reglas: la apertura es manual (jefa de preceptores/admin,
/// desde esta misma pantalla), el cierre normal es automático al pasar
/// `horaLimite`, y `cerradoManual` es solo el cierre anticipado explícito.
class PermisoDiario {
  const PermisoDiario({
    required this.idPermisoDiario,
    required this.fecha,
    required this.usuarioAperturaId,
    required this.horaApertura,
    required this.horaLimite,
    required this.cerradoManual,
  });

  factory PermisoDiario.fromJson(Map<String, dynamic> json) {
    return PermisoDiario(
      idPermisoDiario: json['id_permiso_diario'] as int,
      fecha: json['fecha'] as String,
      usuarioAperturaId: json['usuario_apertura_id'] as int,
      horaApertura: json['hora_apertura'] as String,
      horaLimite: json['hora_limite'] as String,
      cerradoManual: json['cerrado_manual'] as bool,
    );
  }

  /// 'YYYY-MM-DD'.
  final String fecha;

  final int idPermisoDiario;
  final int usuarioAperturaId;

  /// 'YYYY-MM-DD HH:mm:ss' (tal cual la arma `toDateTimeString()` del
  /// lado de Laravel).
  final String horaApertura;

  /// 'HH:mm:ss' — columna TIME de MySQL, sin cast de fecha del lado del
  /// backend, así que llega tal cual.
  final String horaLimite;

  final bool cerradoManual;

  /// 'HH:mm', para mostrar sin los segundos.
  String get horaAperturaCorta =>
      horaApertura.length >= 16 ? horaApertura.substring(11, 16) : horaApertura;

  String get horaLimiteCorta =>
      horaLimite.length >= 5 ? horaLimite.substring(0, 5) : horaLimite;
}

/// Espejo de `PermisosDiariosController::hoy()` — el estado del día para
/// que la pantalla no tenga que "adivinar" nada: si hoy no se abrió
/// ningún permiso todavía, `permiso` es `null` y `abierto` es `false`.
class EstadoPermisoDiario {
  const EstadoPermisoDiario({required this.abierto, required this.permiso});

  factory EstadoPermisoDiario.fromJson(Map<String, dynamic> json) {
    final permisoJson = json['permiso'] as Map<String, dynamic>?;
    return EstadoPermisoDiario(
      abierto: json['abierto'] as bool,
      permiso: permisoJson == null ? null : PermisoDiario.fromJson(permisoJson),
    );
  }

  final bool abierto;
  final PermisoDiario? permiso;
}
